### ISSUE TITLE
RFC: Adding DecimalFloatingPoint abstract type, and the 6 IEEE 754-2008 types to Base

### ISSUE THREAD

**Issue by ScottPJones** [JuliaLang/julia#14241]
It would be useful to have the abstract type `DecimalFloatingPoint`, which is used in the `DecFP.jl` package, and which I also am using in my `decNumber` library based `Dec.jl` package which I hope to submit before too long.
It would also be useful to have defined the 6 types defined by the IEEE 754-2008 standard, which are
32, 64, and 128 bit binary and packed formats (the ranges are identical, the binary format is better suited to software implementations, and is used by the Intel Decimal Floating Point library, and the packed format is better suited to hardware implementations, and is used by IBM's POWER chips and the `decNumber` library.
Note: this is _not_ trying to get decimal floating point into Base, but rather, make it possible to use one or more decimal fp packages, with consistent types.  For example, the Intel library has support for converting between the BID & DPD formats, which I am adding in a PR I am submitting to the `DecFP.jl` package, also the decNumber library has support for arbitrary precision decimal floats, which the Intel library does not, and there is also another pure Julia decimal floating point package that might benefit from this.
This would consist of adding just the following lines somewhere in Base:

``` julia
export DecimalFloatingPoint, Dec32, Dec64, Dec128, D32, D64, D128
abstract DecimalFloatingPoint <: AbstractFloat
bitstype 32 Dec32 <: DecimalFloatingPoint
bitstype 64 Dec64 <: DecimalFloatingPoint
bitstype 128 Dec128 <: DecimalFloatingPoint
bitstype 32 D32 <: DecimalFloatingPoint
bitstype 64 D64 <: DecimalFloatingPoint
bitstype 128 D128 <: DecimalFloatingPoint
```

**Comment by ScottPJones** [comment]
@stevengj @JeffBezanson What are your thoughts on adding *just* this to Base?
If you think it is OK, I'll submit a PR to Julia and another for the `DecFP.jl` package.
I've added the conversion functions already for `DecFP.jl`, but the problems of making it work
with or without also having the `Dec.jl` package I'm doing, and work regardless of order loaded,
made me believe this approach of adding just the types to Base would be necessary.
(I'm also added checked_* functions for `DecFP.jl` as well)

**Comment by JeffBezanson** [comment]
This sounds good; would be great to improve our decimal support.

Should the name be `AbstractDecimal` to be more consistent with `AbstractFloat`? Also the difference between Dec32 and D32 is not clear to me from the names; are these standard names?

**Comment by nalimilan** [comment]
What's the advantage of adding this to Base rather than to a package? I thought the goal was to shrink Base, not to extend it.

**Comment by ScottPJones** [comment]
@JeffBezanson Steven came up with the `DecimalFloatingPoint` and `Dec32`, `Dec64`, `Dec128` names, I just copied that from his nice `DecFP.jl` package.
His package also adds `d32`, `d64`, and `d128` string macros, so I ended up making mine
`D32`, `D64`, and `D128` (which also keeps the string macro names consistent with the type names).
Some bikeshedding on the names would be good, as I'm not happy either with the names.
I'd be happy to use `AbstractDecimal`.  Maybe the type names for the binary format in Steven's package could be changed to `d32`, `d64`, `d128`, and also be consistent with his string macros.

@nalimilan This only adds the types (just 7 lines, plus I've added doc strings ;-) ), *not* any implementation for the types.  Moving forward, I think that might be a good approach for other types, where you might want to have different implementations (say for sparse vectors or arrays), that live in packages, but still be able to write programs that can use a particular type name.  These are also IEEE standard types, I don't think adding *just* the types to Base would be a problem.

**Comment by StefanKarpinski** [comment]
Perhaps it would make sense to have a type parameter on `AbstractFloat` for the base, i.e.:

```jl
abstract AbstractFloat{b} <: Real
typealias BinaryFloatingPoint AbstractFloat{2}
typealias DecimalFloatingPoint AbstractFloat{10}
```

**Comment by ScottPJones** [comment]
I like that idea - would it break anything though, there is a lot of code that has `AbstractFloat`, and it some places it might be assuming that that meant base 2 floating point?
Also, should the two typealias maybe be: `AbstractBinaryFloat` & `AbstractDecimalFloat`, if we want to keep with the whole `Abstract*` scheme?

**Comment by eschnett** [comment]
If there's code relying on a binary representation, it's likely that this code actually expects one of the binary IEEE representations -- `AbstractBinaryIEEEFloat`.

**Comment by J-Sarnoff** [comment]
(imo) We should be assiduous in providing abstract numerical type intermedi-archies that evince established metamathematical intuition.  Nothing good comes of fumbling around for subsumption across adjacent shallow splays of not-entirely-appropriately adjacent abstractions, and vice versa.

Pulling AbstractFloat closer to Real by tying it less immediately to concrete floats is helpful in and of itself --  parameterizing the base for subtypes of AbstractFloat helps helpful fully help.

I have coded, tried, and thrown out some nd-[sub]latticed typings abstractive of number. Julia needs a simple abstract type between Real and AbstractFloat{b} to make more clean numeric type work.
There is a pragmatic reason  https://groups.google.com/forum/#!topic/julia-users/Q4CPfSXNBy4.
And the change from indiscreet Reals into the discrete umbrella of AbstractFloat{b} is a step too steep.

And, should more follow, Floats are an abstract subtype of Surreal numbers, and are no Real subtype.
Irrationals, Integers are Real subtypes. I like typejoin(Real,Float) == Scalar;  the Bool is in your court.

 ```
Surreal
        AbsFloat    
                     AbstractFloat{b}     
                                              BinaryFloat       
                                                                         concrete IEEE types
                                              DecimalFloat
                                                                         concrete IEEE types
Surreal
      Real
           Irrational
                 Integer
```

**Comment by ScottPJones** [comment]
@eschnett Putting IEEE as part of the type (at that level) would not be good, as there are already non-IEEE floating point formats in Julia, such as `BigFloat`.  Also, for decimal fp types, the arbitrary precision types that the `decNumber` package provides are also not IEEE (as well as the pure Julia decimal arithmetic package https://github.com/tinybike/Decimals.jl) This seems to call out for traits, IMO, or at least adding an `isIEEEformat` function.

@J-Sarnoff That seems like something that should be part of a full overhaul of the numeric type hierarchy for Julia, which may be a good thing (not being a mathematician, I'm not sure), but as a programmer, I'd be concerned that it would break a lot of code that expects `Real` to be a superset of `AbstractFloat`.

@StefanKarpinski What about for now, adding:
```julia
abstract AbstractFlt{b} <: Real  # I sure wish there were a way of saying that b was constrained to 2 / 10!
typealias AbstractFloat AbstractFlt{2}
typealias AbstractDecimalFloat AbstractFlt{10}
```
I think that formulation would not break any current code, and would keep `grisu` from trying to get it's hands on the decimal floating point formats.

**Comment by J-Sarnoff** [comment]
@ScottPJones There were several threads in my note.  If one were to limit present adjustment of the type hierarchy to AbstractFloat adjacent abstractions; I would prefer seeing

```julia
abstract RealFloat <: Real
abstract AbstractFloat{b} <: RealFloat
typealias BinaryFloat AbstractFloat{2}
typealias DecimalFloat AbstractFloat{10}
```
This is similar to your suggestion, and equally as non-breaking.  It has the advantage of hewing closer to the IEE754-2008 standard, which worked hard to put base-2 and base-10 numerics on equal footing (hence the new type family namings:  e.g. binary64 and decimal64).

[and briefly noting your concern regarding Surreal, perhaps by then there will exist means to let it work]

**Comment by J-Sarnoff** [comment]
Resubmitting, without other notes
```
abstract AbstractIrrational  <: Real
abstract AbstractFloat{b}    <: AbstractIrrational

typealias BinaryFloat    AbstractFloat{2}
typealias DecimalFloat   AbstractFloat{10}
```
The smallest adjustment of largest benefit to the present number typearchy? AbstractIrrational.

**Comment by ScottPJones** [comment]
Unfortunately, when I tested changing `AbstractFloat` (adding those lines to boot.jl), as suggested by @StefanKarpinski and you, it fails, I'm not sure why:
>Copying in usr/share/doc/julia/examples
    JULIA usr/lib/julia/inference0.ji
ERROR: invalid function object
make[1]: *** [/j/julia/usr/lib/julia/inference0.ji] Error 1
make: *** [julia-inference] Error 2

**Comment by stevengj** [comment]
`AbstractFloat{b}    <: AbstractIrrational` is wrong.  Floating-point numbers are rational.

**Comment by ScottPJones** [comment]
That's what I had thought, but since IANAM, I wasn't sure :grinning:

**Comment by J-Sarnoff** [comment]
Not no over here; though I understand why you thought that.
AbstractIrrational intends something more extensive than Irrational; it is
not the same thing as Irrational.
I used it as shorthand for Rationals+Irrationals (Reals-Transcendentals)

call the intermediating abstract type anything you like



On Thu, Dec 3, 2015 at 10:31 AM, Steven G. Johnson <notifications@github.com

**Comment by stevengj** [comment]
@J-Sarnoff, I don't think it's productive to discuss rearranging the whole numeric hierarchy in this issue.

I like Stefan's
```
abstract AbstractFloat{b} <: Real
typealias BinaryFloatingPoint AbstractFloat{2}
typealias DecimalFloatingPoint AbstractFloat{10}
```
The practical utility of knowing the base is knowing which fractions are represented exactly, and this is independent of whether the format is IEEE.   In that sense, `BigFloat <: BinaryFloatingPoint` is useful information, for example.

In principle, we could have another level of the hierarchy for `IEEEBinaryFloatingPoint`, but I don't think it's that useful — anyone depending on the details of the bit layout will probably want to be working with a concrete type like `Float64` anyway.  We can always add additional intermediate levels to the hierarchy later if a compelling need arises — adding intermediate abstract types is cheap and backwards-compatible.

**Comment by J-Sarnoff** [comment]
ok -- when later comes along, I'll ask about an intermediate level before AbstractFloat

**Comment by StefanKarpinski** [comment]
I'm not convinced that IEEE is a useful abstract type since it's a fixed enumerable set of specific types.

**Comment by StefanKarpinski** [comment]
So basically agreeing with @stevengj here. I'm not sure what the issue with making `AbstractFloat` parametric might be. I'd have to give it a try and work through any bootstrapping issues, which are often kind of nasty. I like the names `BinaryFloat` and `DecimalFloat`. I don't think there's much danger of anyone thinking those are abstract despite the lack of `Abstract` prefix.

**Comment by J-Sarnoff** [comment]
I'm agreeing with that (each one of the IEEE types is specified for concrete realization).

**Comment by ScottPJones** [comment]
@StefanKarpinski No, the `isIEEE` thing I did is more of a trait, easily handled by a few methods, and isn't needed in base either. It was something I found useful in my work, and can be left in my package.
Whatever the nasty bootstrapping issues are, I think those enhancements which you suggested can be saved for another PR.
I've just checked in another attempt - this one does not export anything, and makes the abstract and concrete types (with long names) necessary to be able to have multiple packages that can add different types and formats of binary / decimal floating point formats, without having to be in Base.
As Tony suggested, it will now be the packages responsibility to come up with usable short names.
I didn't use just the base for the parameterization, because there is 1 format for binary floats, and 2 for decimal floats (and it can be more easily extended, to deal with other formats, such as `BigFloat` or `unums` (which might also support both binary and decimal types).
